### ISSUE TITLE
Help: Evaluation and manual testing metric values are different

### ISSUE THREAD

**Issue by AnfimovAleksandr** [manujosephv/pytorch_tabular#577]
# General task background
I'm trying to use Tab Transformer for binary classification using this code (pytorch_tabular part):

data_config = DataConfig(
    target=['target'],  
    continuous_cols=[col for col in full_train.columns if col not in cat_cols and col != 'target'],
    categorical_cols=cat_cols,
)

model_config = TabTransformerConfig(
    task="classification",
    metrics=["f1_score", "average_precision", 'accuracy'],
    metrics_prob_input=[False, False, False]
)

trainer_config = TrainerConfig(
    max_epochs=50,
    early_stopping='valid_loss',
    load_best=False
)

optimizer_config = OptimizerConfig()

tabular_model = TabularModel(
    data_config=data_config,
    model_config=model_config,
    optimizer_config=optimizer_config,
    trainer_config=trainer_config
)

tabular_model.fit(train= full_train)

# Problem
When I evaluate trained model on test dataset, I get different results - evaluate gives higher metric values than manual testing.

## Using model.evaluate()
So evaluate gives this:
eval_dict = tabular_model.evaluate(full_test)
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│       test_accuracy       │     0.87342768907547      │
│  test_average_precision   │     0.931633472442627     │
│       test_f1_score       │     0.87342768907547      │
│         test_loss         │    0.26827090978622437    │
│        test_loss_0        │    0.26827090978622437    │
└───────────────────────────┴───────────────────────────┘

## Using manual testing
from torchmetrics.classification import BinaryF1Score, BinaryAveragePrecision

pred_df = tabular_model.predict(full_test)

f1_metric = BinaryF1Score(threshold=0.5)

f1_metric.update(torch.tensor(pred_df['target_prediction'].to_list()), torch.tensor(y_test.to_list()))
f1_metric.compute().item()

Output: 0.8077003955841064

Even after threshold optimization f1-score is only 0.816, which is much lower than in evaluate().

# Question

Am I missing something and it is intended to work that way? If so can you explain why there is a difference and how I can reliably achieve same metric values?

**Comment by AnfimovAleksandr** [comment]
@dosu, I know that depending on the True/False flag, the estimation of metrics changes. In my case, I use the False flag to calculate the f1-score metric, which means that the metric should be calculated based on predicted values, not probabilities. In theory, these should be the same values that predict() outputs on the same test dataset. However, as the description in the last comment shows, the metrics calculated using predict() are smaller.

**Comment by AnfimovAleksandr** [comment]
@dosu, but even if i use probabilities (which I guess are not reverse engineered) from .predict() to manually make label predictions to calculate f1-score, I still get same value that is lower than in .evaluate(). Also, there is no NaNs or missing values in my data.
I also forgot to mention that my data is unbalanced (33% - 1, and 66% - 0) - maybe evaluate calculates some different f1-score for unbalanced data? If not there is still a difference problem - maybe it's a bug.

**Comment by AnfimovAleksandr** [comment]
@dosu, ok, I understand now. Thank you!
